Update the Dependency Management System (DMS)

wint3ria opened the pull request:
Closes #330 

# Description

Modify the DMS according to #330 , design 3:

- Decouple the DMS configuration from setuptools configuration file (`pyproject.toml`)
- Modify the dependency layer graph (see `graph.png` file). Rename `main` to `dependencies` for naming consistency with `pyproject.toml` 

![graph](https://github.com/eradiate/eradiate/assets/10501833/a3564a66-eaba-4bce-960b-53100745dddf)

- Use `networkx` to handle the layer graph in the conda env generation script
- Import the pip requirement files for the `dependencies` and `recommended` layers in `pyproject.toml`. Only these two sets are thus exposed by the wheel as the `dependencies` list and a `optional-dependencies` list respectively.
- Move the deps definitions for all layers to the `requirements/layered.yml` file. This file now defines the graph and the required packages for each node/layer.
- Move the relevant pip and conda configurations to dedicated folders requirements/pip and requirements/conda
- Allow to output conda deps graph in dot format for debugging

## Layers

**main**: minimal requirements for eradiate to run in dev mode
**depencencies**: dependencies list used by default by setuptools in production packages. Includes the eradiate-mitsuba pkg
**recommended**: convenient optional dependencies included in the production package. Installable through `pip install eradiate[recommended]`
**tests**: required for testing eradiate in dev mode
**docs**: required to compile the docs in dev mode
**dev**: dev setup, includes docs, tests and recommended
**optional**: dev setup including the eradiate-mitsuba pkg for convenience

# Note

This PR removes the previously defined pip and conda lock files, and does not provide any new one. Merging this PR will require to recreate the lock files.

# Checklist

- [x] The code follows the relevant coding guidelines
- [x] The code generates no new warnings
- [x] The code is appropriately documented
- [x] The code is tested to prove its function
- [x] The feature branch is rebased on the current state of the `main` branch
- [x] I updated the change log if relevant
- [x] I give permission that the Eradiate project may redistribute my contributions under the terms of its license

leroyvn reviewed (commented):
Hi @wint3ria, this looks good. The only point I'd like to raise is the deletion of lock files. They are checked in to save the pain of solving the environment during setup (very useful on slow CI/CD machines, *e.g.* on ReadTheDocs). What was your motivation when you decided to remove them?

leroyvn commented:
Regarding your problem with the macOS wheel for Dr.Jit: the wheel you're requesting [seems to exist](https://pypi.org/project/drjit/0.4.1/#files), so it could be that the Poetry solver, used by conda-lock for pip dependencies, just can't find it (pretty embarrassing since Dr.Jit simply doesn't have dependencies). Could worth a report to the conda-lock folks.

wint3ria commented:
> Hi @wint3ria, this looks good. The only point I'd like to raise is the deletion of lock files. They are checked in to save the pain of solving the environment during setup (very useful on slow CI/CD machines, _e.g._ on ReadTheDocs). What was your motivation when you decided to remove them?

I removed them to avoid conflicts on rebase. The purpose of this PR is to *reshape* the way we create these files. I do not mean that lock files are not versioned anymore. The scripts I updated generate lock files into the `requirements/pip` and `requirements/conda` folders instead of putting everything directly in `requirements`. The naming convention is also slightly different. 

Considering all of this, I thought natural to consider previous lock files as outdated, and that we would recreate the files when we rebase/merge this PR. I did not want to keep outdated lock files in the PR, and likely the lock files that I could generate with the new scripts could be outdated after merging. I can indeed provide new lock files as part of this PR if you prefer.

wint3ria commented on Makefile:
```diff
@@ -54,36 +54,53 @@ pip-update: pip-lock pip-init
 	pip install --editable . --no-deps
 
+# Initialise development environment
+conda-init: conda-prepare
+	conda update --file requirements/conda/environment-dev-$(PLATFORM).lock
+	$(MAKE) no-deps-install
```
Yes, I agree. I thought this PR in this sense, extending the behavior you described to allow more lock files to be generated for other types of installation (docs and tests in CI/CD for instance)

wint3ria commented:
> > Hi @wint3ria, this looks good. The only point I'd like to raise is the deletion of lock files. They are checked in to save the pain of solving the environment during setup (very useful on slow CI/CD machines, _e.g._ on ReadTheDocs). What was your motivation when you decided to remove them?
> 
> I removed them to avoid conflicts on rebase. The purpose of this PR is to _reshape_ the way we create these files. I do not mean that lock files are not versioned anymore. The scripts I updated generate lock files into the `requirements/pip` and `requirements/conda` folders instead of putting everything directly in `requirements`. The naming convention is also slightly different.
> 
> Considering all of this, I thought natural to consider previous lock files as outdated, and that we would recreate the files when we rebase/merge this PR. I did not want to keep outdated lock files in the PR, and likely the lock files that I could generate with the new scripts could be outdated after merging. I can indeed provide new lock files as part of this PR if you prefer.

Especially now since we have a problem on osx, we may want for a solution before generating the lock files?

wint3ria commented:
> Regarding your problem with the macOS wheel for Dr.Jit: the wheel you're requesting [seems to exist](https://pypi.org/project/drjit/0.4.1/#files), so it could be that the Poetry solver, used by conda-lock for pip dependencies, just can't find it (pretty embarrassing since Dr.Jit simply doesn't have dependencies and therefore can be installed anywhere). Could be worth a report to conda-lock folks.
> 
> Edit: I posted an issue on the conda-lock issue tracker ([conda/conda-lock#436](https://github.com/conda/conda-lock/issues/436))

After discussion, it was decided to add an option to the conda env generation script to disable the pip deps (and drjit in particular). This is a temporary workaround, until poetry solver is fixed or we have access to a conda package for mitsuba.

wint3ria commented:
As a result of the poetry solver issue, use who may want to use the conda environments `optional` or `dependencies` (from the locks or the env yaml files) will need to perform the installation of the `eradiate-mitsuba` pkg on their own using pip. This is a very minor issue, since these conda environments are only provided for rare corner cases. The `dependencies` layer is used for the wheel main deps list, its purpose isn't to be installed this way with conda. The `optional` layer is only a convenience layer, and most likely won't be used. The new option `-p` for `pip-deps` allow the script to generate conda envs with the pip deps when the solver will be fixed. Complete `dependencies` and `optional` environments can be setup using the pip locks anyway.
